catalog/plugin/catalog-plugin-metacardingest-network/src/main/java/org/codice/ddf/catalog/plugin/metacard/util/AttributeHelper.java

```diff
@@ -0,0 +1,204 @@
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import javax.xml.bind.DatatypeConverter;
+
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.FrameworkUtil;
+import org.osgi.framework.ServiceReference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import ddf.catalog.data.Attribute;
+import ddf.catalog.data.AttributeDescriptor;
+import ddf.catalog.data.Metacard;
+import ddf.catalog.data.MetacardType;
+import ddf.catalog.data.impl.AttributeImpl;
+import ddf.catalog.util.impl.SortedServiceReferenceList;
+
+/**
+ * Support class for working with {@link Attribute}s and {@link Metacard}s.
+ */
+public class AttributeHelper {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(AttributeHelper.class);
+
+    private final Map<String, AttributeDescriptor> systemDescriptors;
+
+    /**
+     * Default constructor for use when the basic attributes are sufficient.
+     */
+    public AttributeHelper() {
+        this.systemDescriptors = new HashMap<>();
+    }
+
+    /**
+     * Constructor given service references to available {@link MetacardType}s.
+     *
+     * @param serviceReferences A list of {@link ServiceReference}s representing exposed
+     *                          {@link MetacardType}s.
+     */
+    public AttributeHelper(SortedServiceReferenceList serviceReferences) {
+        this.systemDescriptors = new HashMap<>();
+
+        BundleContext bundleContext = FrameworkUtil.getBundle(AttributeHelper.class)
+                .getBundleContext();
+        serviceReferences.stream()
+                .map((Function<ServiceReference, Object>) bundleContext::getService)
+                .filter(MetacardType.class::isInstance)
+                .map(MetacardType.class::cast)
+                .filter(Objects::nonNull)
+                .map(MetacardType::getAttributeDescriptors)
+                .forEach(descriptors -> loadUniqueDescriptors(descriptors, systemDescriptors));
+    }
+
+    /**
+     * Iterates over each {@link Metacard} in the {@param metacards} collection and applies each
+     * attribute in the {@param attributeMap} only if the attribute is not already set on the
+     * {@link Metacard}. That is, the attribute must be {@code null} or not in the {@link Metacard}'s
+     * map.
+     *
+     * @param metacards    The list of metacards to attempt to add the attributes to.
+     * @param attributeMap The map of attributes to attempt to add. Attributes already set on any
+     *                     given {@link Metacard} will not be changed.
+     */
+    public void applyNewAttributes(List<Metacard> metacards, Map<String, String> attributeMap) {
+        if (metacards.isEmpty() || attributeMap.isEmpty()) {
+            return;
+        }
+
+        Map<String, AttributeDescriptor> systemAndMetacardDescriptors = new HashMap<>();
+        systemAndMetacardDescriptors.putAll(systemDescriptors);
+        metacards.stream()
```

Comment:
Not exactly ... they have to be registered as a service. It is possible that a custom `MetacardType` with other descriptors not exposed as a service, but added later, could exist. 

Due to thread-safety, I'll need to copy the `systemDescriptors` before unionizing them with the metacard descriptors.